probe-rs-debugger: Add support for stepping at statement level

noppej opened the pull request:
`probe-rs-debugger` now supports stepping at statement level. This includes the ability to 'Step Over', 'Step In', 'Step Out' and 'Run to Cursor'. 

### Some implementation notes: 
- In statement stepping mode, we only step to **'valid' addresses**, i.e. target instructions which are NOT inside a function prologue or epilogue.
- If there is insufficient debug information available to determine a 'valid' target instruction, then the user will be notified and can choose an alternate action.
- Once a target instruction is identified, the debugger will:
  - Temporarily borrow a hardware breakpoint to run to that address.
  - If no hardware breakpoints are available, the debugger will step at instruction level, in a loop,  until it reaches the target instruction.
    - I am not convinced we should allow this last option. If the target instructions are redirected by a hardware interrupt, the debugger may never reach the end of the loop.
- If the user has a VSCode 'Disassembly View' open, then VSCode will automatically use instruction level stepping in all requests, and the 'step over', 'step in', etc. will only step a single instruction and ignore statement level behaviour.

### Additional changes to setting of breakpoints
- The debugger will now use the **'valid' addresses** approach to identifying breakpoint locations requested for a specific source location.
- The debugger will use the requested file/line/column to identify the closest 'valid' statement to step to. 
  - If this location is different from the requested file/line/column, it will notify the DAP client/VSCode of the actual verified position of the breakpoint.
  - If no 'valid' location can be identified, it will notify the user that the breakpoint could not be set as requested.
- In addition to source level breakpoints, the VSCode 'Disassembly View` will allow the setting of breakpoints at ANY target address, and will also allow stepping through instructions which may be outside the definition of 'valid' addresses.

Tiwalun commented:
Just some very quick feedback, could you please move the code into a separate file? Or as much of the code as possible. 

The `probe-rs/src/debug/mod.rs` file is already way too large, and we keep on adding more lines. (I know that I'm very much guilty of this as well...)

noppej commented:
@Tiwalun ... I've resolved the conflicts with master introduced by #1054, so this is ready for review. I do need some help with the right addresses for the tests mentioned in the comment above please.

Tiwalun reviewed (changes requested):
I have some clarifying questions regarding the functions, but other than that looks good to me.

In regards to the tests, I unfortunately didn't keep the original sources for the test files around, so you will have to create some new test cases. We might want to create a new repository for test data, so we can more easily keep track of it.

To create the test cases, I compiled a sample binary and then looked at the disassembly to verify what the proper breakpoint location should be for certain source locations, and then also tested it by actually debugging the program. I don't think there's an easier way to get some test cases for this.

noppej commented on probe-rs/src/debug/mod.rs:
```diff
@@ -528,6 +561,196 @@ impl DebugInfo {
+    /// Validity of halt locations are defined as target instructions that live between the end of the prologue, and the start of the end sequence of a [`gimli::read::LineRow`].
+    ///
+    /// To populate a [`ValidHaltLocations`] struct for a given program_counter.
+    fn get_valid_halt_locations(
```
I have renamed this function and the associated `HaltLocations` struct. I also added more docs to the struct and its fields.

noppej commented on probe-rs/src/debug/mod.rs:
```diff
@@ -321,12 +330,33 @@ type FunctionDieType<'abbrev, 'unit> =
+/// - They point to program row statements where `gimli::read::LineRow::is_stmt()==true`.
+/// - They point to program row statements that are neither inside either the prologue nor inside the epilogue of a function.
+#[derive(Debug)]
+pub struct ValidHaltLocations {
```
I have added some documentation here. I'm hoping this helps to clarify things.

noppej commented on probe-rs/src/debug/mod.rs:
```diff
@@ -1536,6 +1718,176 @@ impl DebugInfo {
+    /// - The best-effort approach is to step a single instruction, and then find the first valid breakpoint address.
+    /// - The worst case is that the user might have to perform the step into action more than once before the debugger properly steps into a function at the specified address.
+    IntoStatement,
+    /// Step to the calling statement, immediately after the current statement.
+    OutOfStatement,
```
Yes, that is what I meant. I updated the comment.

noppej commented:
> I have some clarifying questions regarding the functions, but other than that looks good to me.  

@Tiwalun, I have updated the documentation and renamed the items you suggested. I'd appreciate if you can confirm that this helped to clarify things.

> In regards to the tests, I unfortunately didn't keep the original sources for the test files around, so you will have to create some new test cases. We might want to create a new repository for test data, so we can more easily keep track of it.  

In this case I commented out those test cases and added a TODO item. I'd love to work with you to create something more resilient. The lack of a repeatable test infrastructure adds a lot of manual testing overhead to PR's.

Tiwalun commented:
> n this case I commented out those test cases and added a TODO item. I'd love to work with you to create something more resilient. The lack of a repeatable test infrastructure adds a lot of manual testing overhead to PR's. I can easily justify the time spent to invest in creating a decent infrastructure :)

I would prefer it if this PR retains the same level of tests as before, I don't see why they should be added later. I'm fine with doing it later if there is a good reason for it.

noppej commented:
I agree with keeping the tests, which is why I disabled them, rather than delete them. It is a bit hard to fix them without the original source code. Since there seems to be no alternative, I will work up an example binary and reinstate the tests.

noppej commented:
@Tiwalun I have updated the tests and they are now passing. To prevent future frustration when someone tries to improve this code, I will adjust my probe-rs-debugger-test repo, and reference it from inside these tests, so that others can use and reproduce appropriately.

Tiwalun commented:
Thanks for updating the tests. I think the old binary (`probe-rs/tests/gpio_hal_blinky`) can now be removed. Could you please also add a README or something to indicate the source of the new binary?

I'd like to avoid my old mistakes :)

Tiwalun commented on probe-rs/tests/README.md:
```diff
@@ -0,0 +1,10 @@
+- `inlined_function` 
+  - The source for this binary is unknown. //TODO: Consider re-writing tests against source code in `probe-rs-debugger-test`, and removing the `inlined_function` binary from this repo.
+- `probe-rs-debugger-tests`
+  - This binary was created using the `STM32H745ZITx` feature of the [probe-rs-debugger testing application](https://github.com/probe-rs/probe-rs-debugger-test)
```
Please indicate which exact commit of the repository was used, so that we don't get confused when the repository changes.

Tiwalun commented on probe-rs/tests/source_location.rs:
```diff
@@ -3,23 +3,23 @@ use std::path::{Path, PathBuf};
-    (0x956, 20, ColumnType::Column(12)),
-    (0x962, 21, ColumnType::Column(12)),
-    (0x96a, 22, ColumnType::Column(12)),
+    (0x80006EA, 220, ColumnType::Column(28)),
```
Which source file does this relate to?

Line 220 in the test data repository does not have a column 28:

https://github.com/probe-rs/probe-rs-debugger-test/blob/7803a32ff16c0e74ccc73f67b953b41ea49df1ad/src/main.rs#L220

noppej replied:
Haha ... you're right. I built the binary just before I added support for the other chips. 

Now I'm having doubts if this strategy is going to work, because those tests will likely keep evolving.   

What do you suggest? Go and update the tests and then reference a specific commit in the test README?

Tiwalun replied:
For me, the main goal of the tests is to avoid breaking existing functionality, so I hope that we don't have to change the tests that often. But I agree that the whole setup is a bit cumbersome right now.

noppej replied:
Changes made :-)

Is there some way we can use QEMU or similar to improve this?

Tiwalun commented on probe-rs/tests/README.md:
```diff
@@ -6,5 +6,7 @@ The source code for the tests can be found at locations below. Please note that
-  - This binary was created using the `STM32H745ZITx` feature of the [probe-rs-debugger testing application](https://github.com/probe-rs/probe-rs-debugger-test)
+  - This binary was created using the `STM32H745ZITx` feature of the [probe-rs-debugger testing application](https://github.com/probe-rs/probe-rs-debugger-test), and specific file versions as below:
+    - `main.rs` : https://github.com/probe-rs/probe-rs-debugger-test/blob/565132e3002860b1535908fbe7bf717188345ee7/src/main.rs
+    - `Cargo.lock`: https://github.com/probe-rs/probe-rs-debugger-test/blob/259dce3a8677c8bfe8382328d142c9a8995a3128/Cargo.lock
```
This seems to point to a different commit compared to the `main.rs` file. Just mentioning a single commit for all files is probably easier.

It would also be good to indicate if the binary was compiled in release or debug mode, maybe just mention the correct command?

Tiwalun replied:
If I want to reproduce the binary, I want to be able to do something like this:

- git checkout <commit id>
- cargo build --target xxx --features xxx --locked

But right now this is not easily done, since there are two separate commits mentioned, and I first have to figure out the necessary arguments to cargo. The `[profile.dev]` section changes the settings for a debug build, but that doesn't mean that I can't just use `--release` to perform a release build, so it would be nice to clearly specify how the binary is built.

noppej replied:
Done

noppej commented:
bors r+
